src/dns.rs

```diff
@@ -0,0 +1,97 @@
+mod local;
+mod default;
+mod cache;
+
+use crate::types::Result;
+use std::net::IpAddr;
+use std::time;
+
+/// A single IP entry in the DNS cache. Each ip entry has an expiration time (based on the TTL)
+/// and a last-used timestamp. When the entry is expired, the entry is found to be stale and should
+/// be refreshed by the resolver. When expired == 0, the entry is considered to be a local override
+/// and should not be refreshed by the resolver.
+/// @todo: Does that information belong here? Or should it be in the cache layer?
+struct IpEntry {
+    /// Actual IP address (either ipv4 or ipv6)
+    ip: IpAddr,
+    /// Epoch when ip address expires
+    expired: u64,
+    /// Epoch when this entry was last used
+    last_used: u64,
+}
+
+/// A DNS entry is a simple domain name to IP address mapping
+#[derive(Default)]
+struct DNSEntry {
+    // domain name
+    domain_name: String,
+    // List of IPv4 addresses for this domain name
+    ipv4: Vec<IpEntry>,
+    // List of IPv6 addresses for this domain name
+    ipv6: Vec<IpEntry>,
+    // Round robin pointer in case we need to round robin
+    rr_ipv4_ptr: usize,
+    // Round robin pointer in case we need to round robin
+    rr_ipv6_ptr: usize,
```

Comment:
Should this be an enum?  Also, curious why we aren't making use of a simple wrapper around an existing Rust library.  This looks more low-level than I was anticipating.

Reply:
We are using hickory dns in the remote resolver. I'll draw up a diagram to give you an idea of the architecture (its not too complicated)